packages/vm/src/context.rs

```diff
@@ -341,33 +348,177 @@ pub fn with_querier_from_context<S: Storage, Q: Querier, F: FnMut(&Q)>(ctx: &Ctx
     b.querier = querier;
 }
 
-pub fn take_storage<S: Storage, Q: Querier>(ctx: &Ctx) -> Option<S> {
+/// take_context_data will return the original storage and querier, and closes any remaining
+/// iterators. This is meant to be called when recycling the instance
+pub(crate) fn take_context_data<S: Storage, Q: Querier>(ctx: &Ctx) -> (Option<S>, Option<Q>) {
     let b = unsafe { get_data::<S, Q>(ctx.data) };
-    // we do this to avoid cleanup
     let mut b = mem::ManuallyDrop::new(b);
-    b.storage.take()
+    // free out the iterator as this finalizes the instance
+    free_iterator(&mut b);
+    (b.storage.take(), b.querier.take())
 }
 
-pub fn leave_storage<S: Storage, Q: Querier>(ctx: &Ctx, storage: Option<S>) {
+/// leave_context_data sets the original storage and querier. These must both be set.
+/// Should be followed by exactly one call to take_context_data when the instance is finished.
+pub(crate) fn leave_context_data<S: Storage, Q: Querier>(ctx: &Ctx, storage: S, querier: Q) {
     let b = unsafe { get_data::<S, Q>(ctx.data) };
-    // we do this to avoid cleanup
-    let mut b = mem::ManuallyDrop::new(b);
-    b.storage = storage;
+    let mut b = mem::ManuallyDrop::new(b); // we do this to avoid cleanup
+    b.storage = Some(storage);
+    b.querier = Some(querier);
 }
 
-pub fn take_context_data<S: Storage, Q: Querier>(ctx: &Ctx) -> (Option<S>, Option<Q>) {
-    let b = unsafe { get_data::<S, Q>(ctx.data) };
-    let mut b = mem::ManuallyDrop::new(b);
-    (b.storage.take(), b.querier.take())
-}
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::backends::compile;
+    use cosmwasm_std::testing::{MockQuerier, MockStorage};
+    use cosmwasm_std::{coin, ReadonlyStorage};
+    use wasmer_runtime_core::{imports, instance::Instance, typed_func::Func};
 
-pub fn leave_context_data<S: Storage, Q: Querier>(
-    ctx: &Ctx,
-    storage: Option<S>,
-    querier: Option<Q>,
-) {
-    let b = unsafe { get_data::<S, Q>(ctx.data) };
-    let mut b = mem::ManuallyDrop::new(b); // we do this to avoid cleanup
-    b.storage = storage;
-    b.querier = querier;
+    #[cfg(feature = "iterator")]
+    use super::iter_support::with_iterator_from_context;
+
+    static CONTRACT: &[u8] = include_bytes!("../testdata/contract.wasm");
+
+    // shorthand for function generics below
+    type S = MockStorage;
+    type Q = MockQuerier;
+
+    // prepared data
+    static INIT_KEY: &[u8] = b"foo";
+    static INIT_VALUE: &[u8] = b"bar";
+    // this account has some coins
+    static INIT_ADDR: &str = "someone";
+    static INIT_AMOUNT: &str = "500";
+    static INIT_DENOM: &str = "TOKEN";
+
+    fn make_instance() -> Instance {
+        let module = compile(&CONTRACT).unwrap();
+        let mut import_obj =
+            imports! { || { setup_context::<MockStorage, MockQuerier>() }, "env" => {}, };
+        // we need stubs for all required imports
+        import_obj.extend(imports! {
```

Comment:
Here we can simplify code to just one call of `imports!` with immutable `let import_obj`